spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -372,6 +375,15 @@ private void setEvaluatingSpelQueryContext() {
 			}
 		}
 
+		private Object convertEntitiesToKeys(Object o, DatastoreTemplate datastoreTemplate, DatastoreMappingContext datastoreMappingContext) {
+			Entity entity = AnnotationUtils.findAnnotation(o.getClass(), Entity.class);
+			if (entity != null) {
+				return datastoreTemplate.getObjectToKeyFactory()
+						.getKeyFromObject(o, datastoreMappingContext.getPersistentEntity(o.getClass()));
```

Comment:
Can you do this logic in `AbstractDatastoreQuery.convertCollectionParamToCompatibleArray` so that it applies to all types of queries?

Reply:
This method is used to convert collection-like parameters to arrays of Datastore Values, so I don't think it's a good fit

Reply:
I see. I still think that doing this conversion in `ParsedQueryWithTagsAndValues()` is not the best place. The rest of the parameter conversion is happening in `GqlDatastoreQuery.bindArgsToGqlQuery()`. Can you do it there?
Maybe it can even be done in `ReadWriteConversions.convertOnWriteSignle()`?